File: bun/trees.js
// trees.js: regression / decision trees over a Data
// (c) 2026 Tim Menzies, timm@ieee.org, MIT license

import {
  the, loadDoc, say, Num, add, adds, clone, mid, spread, disty, o, mn, flr
} from "./core.js";

loadDoc(`--learn.leaf=3   min examples per leaf`);

function treeCuts(col, rows) {
  let vs = [];
  for (let r of rows) if (r[col.at] !== "?") vs.push(r[col.at]);
  if (!vs.length) return [];
  if (col.type === "Sym") return [...new Set(vs)];
  vs.sort((a,b) => a-b);
  return [vs[flr(vs.length / 2)]]; }

function treeSplit(data, col, cut, rows) {
  let lr=[], rr=[], ln=Num(), rn=Num();
  for (let row of rows) {
    let v  = row[col.at];
    let go = v === "?" || (col.type === "Sym" ? v === cut : v <= cut);
    (go ? lr : rr).push(row);
    add(go ? ln : rn, disty(data, row)); }
  return {s: ln.n*spread(ln) + rn.n*spread(rn), col, cut, lr, rr}; }

function treeGrow(data, rows) {
  let tree = { type:"Tree", d: clone(data, rows),
               ynum: adds(rows.map(r => disty(data, r)), Num()),
               col: null, cut: 0, left: null, right: null };
  if (rows.length >= 2 * the.learn.leaf) {
    let valid = [];
    for (let col of tree.d.cols.xs)
      for (let cut of treeCuts(col, rows)) {
        let sp = treeSplit(data, col, cut, rows);
        if (mn(sp.lr.length, sp.rr.length) >= the.learn.leaf) valid.push(sp); }
    if (valid.length) {
      let best = valid.reduce((a,b) => b.s < a.s ? b : a);
      tree.col = best.col; tree.cut = best.cut;
      tree.left  = treeGrow(data, best.lr);
      tree.right = treeGrow(data, best.rr); } }
  return tree; }

function treeLeaf(tree, row) {
  if (!tree.left) return tree;
  let v  = row[tree.col.at];
  let go = v !== "?" && (tree.col.type === "Num" ? v <= tree.cut
                                                 : v === tree.cut);
  return treeLeaf(go ? tree.left : tree.right, row); }

function* treeNodes(tree, lvl=0, col=null, op="", cut=null) {
  yield [tree, lvl, col, op, cut];
  if (tree.col) {
    let ops  = tree.col.type === "Num" ? ["<=",">"] : ["==","!="];
    let kids = [[tree.left, ops[0]], [tree.right, ops[1]]];
    kids.sort((a,b) => mid(a[0].ynum) - mid(b[0].ynum));
    for (let [k, txt] of kids)
      if (k) yield* treeNodes(k, lvl+1, tree.col, txt, tree.cut); } }

function treeShow(tree) {
  for (let [t1, lvl, col, op, cut] of treeNodes(tree)) {
    let p = (lvl > 0 ? "|   ".repeat(lvl-1) : "")
          + (col ? `${col.txt} ${op} ${o(cut)}` : "");
    let g = Object.fromEntries(t1.d.cols.ys.map(c => [c.txt, mid(c)]));
    say(`${p.padEnd(the.show.show)},${o(mid(t1.ynum)).padStart(4)}`
      + ` ,(${String(t1.ynum.n).padStart(3)}), ${o(g)}`); } }

export { treeCuts, treeSplit, treeGrow, treeLeaf, treeNodes, treeShow };

File: bun/trees.py
#!/usr/bin/env python3 -B
"""
trees.py: regression / decision trees over a Data.

Local options:
    --learn.leaf=3   min examples per leaf
"""
from core import *
loadDoc(__doc__)

class Tree:
  def __init__(i, data, rows):
    i.d = clone(data, rows)
    i.ynum = adds((disty(data, r) for r in rows), Num())
    i.col, i.cut = None, 0
    i.left = i.right = None

def treeCuts(col, rows):
  vs = [r[col.at] for r in rows if r[col.at] != "?"]
  if not vs: return []
  return set(vs) if Sym == type(col) else [sorted(vs)[len(vs)//2]]

def treeSplit(data, col, cut, rows):
  l_rows, r_rows, l_num, r_num = [], [], Num(), Num()
  for row in rows:
    v  = row[col.at]
    go = v == "?" or (v == cut if Sym == type(col) else v <= cut)
    (l_rows if go else r_rows).append(row)
    add(l_num if go else r_num, disty(data, row))
  s = l_num.n*spread(l_num) + r_num.n*spread(r_num)
  return s, col, cut, l_rows, r_rows

def treeGrow(data, rows):
  tree = Tree(data, rows)
  if len(rows) >= 2 * the.learn.leaf:
    splits = (treeSplit(data, c, cut, rows)
              for c in tree.d.cols.xs
              for cut in treeCuts(c, rows))
    if valid := [s for s in splits
                 if min(len(s[3]), len(s[4])) >= the.learn.leaf]:
      _, tree.col, tree.cut, left, right = min(valid, key=lambda x: x[0])
      tree.left  = treeGrow(data, left)
      tree.right = treeGrow(data, right)
  return tree

def treeLeaf(tree, row):
  if not tree.left: return tree
  v  = row[tree.col.at]
  go = v != "?" and (v <= tree.cut if Num == type(tree.col)
                                   else v == tree.cut)
  return treeLeaf(tree.left if go else tree.right, row)

def treeNodes(tree, lvl=0, col=None, op="", cut=None):
  yield tree, lvl, col, op, cut
  if tree.col:
    ops = ("<=", ">") if Num == type(tree.col) else ("==", "!=")
    kids = sorted([(tree.left, ops[0]), (tree.right, ops[1])],
                  key=lambda z: mid(z[0].ynum))
    for k, txt in kids:
      if k: yield from treeNodes(k, lvl+1, tree.col, txt, tree.cut)

def treeShow(tree):
  for t1, lvl, col, op, cut in treeNodes(tree):
    p = f"{col.txt} {op} {o(cut)}" if col else ""
    if lvl > 0: p = "|   " * (lvl - 1) + p
    g = {c.txt: mid(c) for c in t1.d.cols.ys}
    print(f"{p:<{the.show.show}}"
          f",{o(mid(t1.ynum)):>4}"
          f" ,({t1.ynum.n:3}), {o(g)}")

__all__ = ["Tree", "treeCuts", "treeSplit", "treeGrow",
           "treeLeaf", "treeNodes", "treeShow"]
